### PR TITLE
Issues interface

### PR DESCRIPTION
**Pull request by tiginamaria**
We discussed a problem, that we should not run script for issues preparing and have an ability to parse, filter and dump 'raw' issue reports. So in this PR, I added the common interface for all issues. It allows unifying parsing/analysis of code quality reports from different analyzers.

### CODE REVIEW

File: analysis/src/python/evaluation/tools/hyperstyle/model/report.py [hunk 1]
```diff
@@ -38,17 +57,25 @@ def to_str(self) -> str:
...
-class HyperstyleReport(QualityReport):
+@dataclass(frozen=True, eq=True)
+class HyperstyleReport(QualityReport, BaseReport):
     issues: List[HyperstyleIssue]
 
+    def get_issues(self) -> List[HyperstyleIssue]:
+        return self.issues
+
+    def filter_issues(self, predicate: Callable[[BaseIssue], bool]) -> 'HyperstyleReport':
+        # TODO: recalculate quality after filtering
```
Comment (nbirillo) [inline analysis/src/python/evaluation/tools/hyperstyle/model/report.py:68]: How are you going to recalculate it?

File: analysis/src/python/data_analysis/preprocessing/preprocess_issues.py [hunk 2]
```diff
@@ -17,9 +17,9 @@ def get_issues_info(df_submissions: pd.DataFrame, issues_column: str) -> pd.Data
     issues_info = {}
 
     def get_info(str_report: str):
-        report = AnalysisReport.from_json(str_report)
-        for issue in report.issues:
-            issues_info[issue.name] = issue.category
+        report = parse_str_report(str_report, issues_column)
+        for issue in report.get_issues():
```
[the remark this answers is not in the document]
Comment (nbirillo) [reply analysis/src/python/data_analysis/preprocessing/preprocess_issues.py:21]: I mean we can mark them as @property

File: analysis/src/python/evaluation/tools/model/report.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    """ All code quality reports which a going to be analyzed should implement this abstract class. """
+
+    def has_issue(self, name: str):
+        for issue in self.get_issues():
+            if issue.get_name() == name:
+                return True
+        return False
+
+    @abstractmethod
+    def get_issues(self) -> List[BaseIssue]:
```
Comment (nbirillo) [inline analysis/src/python/evaluation/tools/model/report.py:64]: And here - maybe is it better to use @property?

**Review by nbirillo — CHANGES_REQUESTED** [review]
Great! Only several comments

File: analysis/src/python/data_analysis/preprocessing/preprocess_issues.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tiginamaria) [reply analysis/src/python/data_analysis/preprocessing/preprocess_issues.py:21]: @property is what I need! I have not heard about it, will add it, thank you a lot!

File: analysis/src/python/evaluation/tools/hyperstyle/model/report.py (hunk 1 above)
Comment (tiginamaria) [reply analysis/src/python/evaluation/tools/hyperstyle/model/report.py:68]: It is a tip for the future if we need it. When we filter template issues, for example, the quality can change, but for now I use one from the original report =( It is ok for now but if we want to analyze quality, we need to recalculate it

**Comment by tiginamaria** [comment]
We decided that @property is hard to use with dataclasses so getter is ok, if we do not want to write serialization/deserialization manually.

**Review by nbirillo — APPROVED** [review]

**Pull request merged**